File: server/protocol/protocol.py
ACK_OK = b"ACK|OK"
ACK_FAIL = b"ACK|FAIL"
MAX_PAYLOAD_SIZE = 8192

# read_exact lee exactamente 'size' bytes del socket.
# Si la conexión se cierra antes de completar la lectura, tira un error.
def read_exact(sock, size):
    data = bytearray()
    while len(data) < size:
        chunk = sock.recv(size - len(data))
        if not chunk:
            raise OSError("connection closed before reading expected bytes")
        data.extend(chunk)
    return bytes(data)

# read_frame lee un frame completo del socket con formato:
# [2 bytes de tamaño (big-endian)][payload]
def read_frame(sock):
    # Lee el header de 2 bytes que indica el tamaño
    header = read_exact(sock, 2)
    payload_size = int.from_bytes(header, byteorder="big")
    if payload_size > MAX_PAYLOAD_SIZE:
            raise ValueError(f"Payload recibido excede el límite: {payload_size}")
    
    # Lee el payload completo según el tamaño indicado
    return read_exact(sock, payload_size)

# write_frame escribe un frame en el socket con formato:
# [2 bytes de tamaño (big-endian)][payload]
# Usa sendall para asegurarse de que se envíen todos los bytes.
def write_frame(sock, payload):
    # Valida que el payload no exceda MAX_PAYLOAD_SIZE bytes
    if len(payload) > MAX_PAYLOAD_SIZE:
        raise ValueError(f"payload too large: {len(payload)} bytes (max: {MAX_PAYLOAD_SIZE})")
    
    # Arma el header con el tamaño en big-endian y envía el header + payload
    header = len(payload).to_bytes(2, byteorder="big")
    sock.sendall(header + payload)

File: client/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"io"
)

const (
	ExpectedACK     = "ACK|OK"
	ExpectedNACK    = "ACK|FAIL"
	ExpectedACKWait = "ACK|WAIT"
)

// MaxPayloadSize es el tamaño máximo de payload en bytes.
// Es configurable desde config.yaml (protocol.maxPayloadSize).
// Default: 8192 (8KB)
var MaxPayloadSize = 8192

// SetMaxPayloadSize permite configurar el tamaño máximo de payload.
func SetMaxPayloadSize(size int) {
	if size > 0 {
		MaxPayloadSize = size
	}
}

// WriteFrame escribe un frame en conexión con el formato:
// 2 bytes size (uint16 big-endian)][payload]
// Maneja posibles short-writes iterando hasta que se escriban todos los bytes.
func WriteFrame(conn io.Writer, payload []byte) error {
	// Valida que el tamaño del payload no exceda 8192 bytes
	if len(payload) > MaxPayloadSize {
		return fmt.Errorf("payload too large: %d bytes (max: %d)", len(payload), MaxPayloadSize)
	}

	// Crea el frame: [2 bytes header][payload]
	frame := make([]byte, 2+len(payload))

	// Escribe el tamaño en formato big-endian
	binary.BigEndian.PutUint16(frame[0:2], uint16(len(payload)))

	// Copia el payload
	copy(frame[2:], payload)

	// Escribe todo el frame manejando short-writes
	totalWritten := 0
	for totalWritten < len(frame) {
		n, err := conn.Write(frame[totalWritten:])
		if err != nil {
			return err
		}
		totalWritten += n
	}

	return nil
}

// ReadFrame lee un frame completo desde la conexión.
// Maneja posibles lecturas short-reads leyendo exactamente los bytes necesarios.
func ReadFrame(conn io.Reader) ([]byte, error) {
	// Lee el header de tamaño (2 bytes)
	sizeBuf := make([]byte, 2)
	totalRead := 0
	for totalRead < 2 {
		n, err := conn.Read(sizeBuf[totalRead:])
		if err != nil {
			return nil, fmt.Errorf("error reading frame size: %w", err)
		}
		totalRead += n
	}

	// Parsea el tamaño desde big-endian
	size := binary.BigEndian.Uint16(sizeBuf)

	// Lee el payload
	payload := make([]byte, size)
	totalRead = 0
	for totalRead < int(size) {
		n, err := conn.Read(payload[totalRead:])
		if err != nil {
			return nil, fmt.Errorf("error reading frame payload: %w", err)
		}
		totalRead += n
	}

	return payload, nil
}
